nemo/core/classes/modelPT.py

```diff
@@ -1400,3 +1403,75 @@ def update_save_restore_connector(cls, save_restore_connector):
             cls._save_restore_connector = save_restore_connector
         else:
             setattr(cls, '_save_restore_connector', save_restore_connector)
+
+    def _setup_nsys_profiling(self):
+        """ Enables nsys profiling
+            To use, add the following optoins to the model config:
+            ## Nsys profiling options
+            nsys_profile: False
+                start_step: 10  # Global batch to start profiling
+                end_step: 10 # Global batch to end profiling
+                ranks: [0] # Global rank IDs to profile
+                gen_shape: False # Generate model and kernel details including input shapes
+            And then wrap the model training script with:
+            nsys profile -s none -o <profile filepath>  -t cuda,nvtx --force-overwrite true --capture-range=cudaProfilerApi --capture-range-end=stop python ./examples/...
+            See more options at: https://docs.nvidia.com/nsight-systems/UserGuide/index.html#cli-profiling
+        """
+        if self.cfg.get('nsys_profile', None):
+            if self.cfg.nsys_profile.get('enabled', False):
+                # Nsys profiling options
+                self._nsys_profile_start_step = self.cfg.nsys_profile.get('start_step', 0)
+                self._nsys_profile_end_step = self.cfg.nsys_profile.get('end_step', 0)
+                self._nsys_profile_ranks = self.cfg.nsys_profile.get('ranks', [0])
+                self._nsys_profile_gen_shape = self.cfg.nsys_profile.get('gen_shape', False)
+
+                if type(self._nsys_profile_start_step) == int:
+                    logging.info(f'Nsys profiling setup with start_step: {self._nsys_profile_start_step}')
+                else:
+                    raise ValueError(
+                        f'Nsys start_step must be of type int. Found: {type(self._nsys_profile_start_step)}'
+                    )
+
+                if type(self._nsys_profile_end_step) == int:
+                    logging.info(f'Nsys profiling setup with end_step: {self._nsys_profile_end_step}')
+                else:
+                    raise ValueError(f'Nsys end_step must be of type int. Found: {type(self._nsys_profile_end_step)}')
+
+                if self._nsys_profile_end_step >= self._nsys_profile_start_step:
+                    pass
+                else:
+                    raise ValueError(f'Nsys end_step must be greater than or equal to nsys start_step')
+
+    def on_train_batch_start(self, batch: Any, batch_idx: int, unused: int = 0) -> Optional[int]:
+        """ PyTorch Lightning hook: 
+            https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#on-train-batch-start
+            We use it here to enable nsys profiling.
+        """
+
+        if self.device.type == 'cuda':
+            if self.cfg.get('nsys_profile', None):
```

Comment:
Hmm, wont this be a performance penalty to access the OmegaConf object every single batch? How about 

if hasattr(self, '_nsys_profile_enabled') and self._nsys_profile_enabled: ... 
where `self._nsys_profile_enabled` is set inside of the `_setup_nsys_profiling()` method.

Reply:
Great catch. Done.